src/writer.rs

```diff
@@ -563,263 +728,307 @@ impl<D: Distance> Writer<D> {
+            .delete_range(wtxn, &(Key::tree(self.index, 0)..=Key::tree(self.index, ItemId::MAX)))?;
+        let mut roots = Vec::new();
+        if !item_indices.is_empty() {
+            // if we have more than 0 elements we need to create a descendant node
 
-        if thread_roots.is_empty() {
-            // we may have too many nodes
-            tracing::debug!("Deleting the extraneous trees if there is some...");
-            (options.progress)(WriterProgress { main: MainStep::DeleteExtraneousTrees, sub: None });
-            self.delete_extra_trees(
+            self.database.put(
                 wtxn,
-                options,
-                &mut roots,
-                options.n_trees,
-                concurrent_node_ids.used(),
-                n_items,
+                &Key::tree(self.index, 0),
+                &Node::Descendants(Descendants { descendants: Cow::Borrowed(&item_indices) }),
             )?;
-        } else {
-            roots.append(&mut thread_roots);
+            roots.push(0);
         }
-
+        options.cancelled()?;
         tracing::debug!("write the metadata...");
-        (options.progress)(WriterProgress { main: MainStep::WriteTheMetadata, sub: None });
         let metadata = Metadata {
             dimensions: self.dimensions.try_into().unwrap(),
             items: item_indices,
             roots: ItemIds::from_slice(&roots),
             distance: D::name(),
         };
-        match self.database.remap_data_type::<MetadataCodec>().put(
+        self.database.remap_data_type::<MetadataCodec>().put(
             wtxn,
             &Key::metadata(self.index),
             &metadata,
-        ) {
-            Ok(_) => Ok(()),
-            Err(e) => Err(e.into()),
+        )?;
+        tracing::debug!("write the version...");
+        let version = Version {
+            major: env!("CARGO_PKG_VERSION_MAJOR").parse().unwrap(),
+            minor: env!("CARGO_PKG_VERSION_MINOR").parse().unwrap(),
+            patch: env!("CARGO_PKG_VERSION_PATCH").parse().unwrap(),
+        };
+        self.database.remap_data_type::<VersionCodec>().put(
+            wtxn,
+            &Key::version(self.index),
+            &version,
+        )?;
+        Ok(())
+    }
+
+    fn pre_process_items(&self, wtxn: &mut RwTxn, options: &BuildOption) -> Result<(), Error> {
+        tracing::debug!("started preprocessing the items...");
+        (options.progress)(WriterProgress { main: MainStep::PreProcessingTheItems, sub: None });
+        options.cancelled()?;
+        D::preprocess(wtxn, |wtxn| {
+            Ok(self
+                .database
+                .remap_key_type::<PrefixCodec>()
+                .prefix_iter_mut(wtxn, &Prefix::item(self.index))?
+                .remap_key_type::<KeyCodec>())
+        })?;
+        Ok(())
+    }
+
+    // TODO: Should we return the list of empty descendants to double check at the very end of the indexing process if
+    // they're still empty and should be removed?
```

Comment:
I don't know, you tell me 😂 Do you plan to talk about this with me or to open an issue or it's ok?

Reply:
yeah I don’t I was just thinking out loud about what we’ll need to paralellize but we can get rids of this comment